# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources in the README.md file
- Happy learning!

# Challenge 1 - Accessing our database
Create a connection to access the sakila database with the ip, user and password provided in class. Take a look at the data. Do some joins in order to have more elaborated tables in a dataframe (you can try first the query in the DBMS and then use it in your notebook).


In [10]:
# your code here
import pandas as pd
from sqlalchemy import create_engine
import pymysql

driver = "mysql+pymysql"
user = "root"
password = "*****"
ip = "localhost"
database = "sakila"

connection_string = f"{driver}://{user}:{password}@{ip}/{database}"

engine = create_engine(connection_string)
query1 = """ SELECT * FROM actor
"""
query2 = """ SELECT * FROM film_actor
"""
query3 = """ SELECT * FROM film
"""

In [1]:
actor = pd.read_sql(query1, engine)
actor.head()

NameError: name 'pd' is not defined

In [12]:
actor_film = pd.read_sql(query2, engine)
actor_film.head()

actor_id  film_id         last_update
0         1        1 2006-02-15 05:05:03
1         1       23 2006-02-15 05:05:03
2         1       25 2006-02-15 05:05:03
3         1      106 2006-02-15 05:05:03
4         1      140 2006-02-15 05:05:03

In [13]:
film= pd.read_sql(query3, engine)
film.head()

film_id             title  \
0        1  ACADEMY DINOSAUR   
1        2    ACE GOLDFINGER   
2        3  ADAPTATION HOLES   
3        4  AFFAIR PREJUDICE   
4        5       AFRICAN EGG   

                                         description  release_year  \
0  A Epic Drama of a Feminist And a Mad Scientist...          2006   
1  A Astounding Epistle of a Database Administrat...          2006   
2  A Astounding Reflection of a Lumberjack And a ...          2006   
3  A Fanciful Documentary of a Frisbee And a Lumb...          2006   
4  A Fast-Paced Documentary of a Pastry Chef And ...          2006   

   language_id original_language_id  rental_duration  rental_rate  length  \
0            1                 None                6         0.99      86   
1            1                 None                3         4.99      48   
2            1                 None                7         2.99      50   
3            1                 None                5         2.99     117   
4            1                 None                6         2.99     130   

   replacement_cost rating                  special_features  \
0             20.99     PG  Deleted Scenes,Behind the Scenes   
1             12.99      G           Trailers,Deleted Scenes   
2             18.99  NC-17           Trailers,Deleted Scenes   
3             26.99      G    Commentaries,Behind the Scenes   
4             22.99      G                    Deleted Scenes   

          last_update  
0 2006-02-15 05:03:42  
1 2006-02-15 05:03:42  
2 2006-02-15 05:03:42  
3 2006-02-15 05:03:42  
4 2006-02-15 05:03:42

In [24]:
actor_film_names = actor_film.merge(actor, how= "inner", on = "actor_id")
actor_film_names.head()

actor_id  film_id       last_update_x first_name last_name  \
0         1        1 2006-02-15 05:05:03   PENELOPE   GUINESS   
1         1       23 2006-02-15 05:05:03   PENELOPE   GUINESS   
2         1       25 2006-02-15 05:05:03   PENELOPE   GUINESS   
3         1      106 2006-02-15 05:05:03   PENELOPE   GUINESS   
4         1      140 2006-02-15 05:05:03   PENELOPE   GUINESS   

        last_update_y  
0 2006-02-15 04:34:33  
1 2006-02-15 04:34:33  
2 2006-02-15 04:34:33  
3 2006-02-15 04:34:33  
4 2006-02-15 04:34:33

In [33]:
query4 = """ SELECT title, release_year, f.film_id, first_name AS actor_name, last_name AS actor_2_name
FROM actor as a
LEFT JOIN film_actor as af
ON a.actor_id = af.actor_id

LEFT JOIN film as f
ON af.film_id = f.film_id



"""
new_af= pd.read_sql(query4, engine)
new_af.head()

title  release_year  film_id actor_name actor_2_name
0       ACADEMY DINOSAUR          2006        1   PENELOPE      GUINESS
1   ANACONDA CONFESSIONS          2006       23   PENELOPE      GUINESS
2            ANGELS LIFE          2006       25   PENELOPE      GUINESS
3  BULWORTH COMMANDMENTS          2006      106   PENELOPE      GUINESS
4          CHEAPER CLYDE          2006      140   PENELOPE      GUINESS

# Challenge 2 - Working with JSON files

Import the pandas library.

In [34]:
# your code here
import pandas as pd

In the next cell, load the data in `nasa.json` in the `data` folder and load it into a pandas dataframe. Name the dataframe `nasa`.

In [35]:
# your code here
nasa = pd.read_json("../data/nasa.json")

Now that we have loaded the data, let's examine it using the `head()` function.

In [36]:
# your code here
nasa.head()

:@computed_region_cbhk_fwbd  :@computed_region_nnqa_25f4  fall  \
0                          NaN                          NaN  Fell   
1                          NaN                          NaN  Fell   
2                          NaN                          NaN  Fell   
3                          NaN                          NaN  Fell   
4                          NaN                          NaN  Fell   

                                         geolocation   id      mass      name  \
0  {'type': 'Point', 'coordinates': [6.08333, 50....    1      21.0    Aachen   
1  {'type': 'Point', 'coordinates': [10.23333, 56...    2     720.0    Aarhus   
2  {'type': 'Point', 'coordinates': [-113, 54.216...    6  107000.0      Abee   
3  {'type': 'Point', 'coordinates': [-99.9, 16.88...   10    1914.0  Acapulco   
4  {'type': 'Point', 'coordinates': [-64.95, -33....  370     780.0   Achiras   

  nametype     recclass    reclat    reclong                     year  
0    Valid           L5  50.77500    6.08333  1880-01-01T00:00:00.000  
1    Valid           H6  56.18333   10.23333  1951-01-01T00:00:00.000  
2    Valid          EH4  54.21667 -113.00000  1952-01-01T00:00:00.000  
3    Valid  Acapulcoite  16.88333  -99.90000  1976-01-01T00:00:00.000  
4    Valid           L6 -33.16667  -64.95000  1902-01-01T00:00:00.000

#### The `value_counts()` function is commonly used in pandas to find the frequency of every value in a column.

In the cell below, use the `value_counts()` function to determine the frequency of all types of asteroid landings by applying the function to the `fall` column.

In [37]:
# your code here
nasa["fall"].value_counts()

Fell     996
Found      4
Name: fall, dtype: int64

Finally, let's save the dataframe as a json file again. Save the dataframe using the `orient=records` argument and name the file `nasa-output.json`. Remember to save the file inside the `data` folder.

In [38]:
# your code here
nasa["fall"].value_counts().to_json("../data/nasa-output.json", orient='records')

# Challenge 2 - Working with CSV and Other Separated Files

CSV files are more commonly used as dataframes. In the cell below, load the file from the URL provided using the `read_csv()` function in pandas. Starting version 0.19 of pandas, you can load a CSV file into a dataframe directly from a URL without having to load the file first and then transform it. The dataset we will be using contains information about NASA shuttles.

In the cell below, we define the column names and the URL of the data. Following this cell, read the tst file to a variable called `shuttle`. Since the file does not contain the column names, you must add them yourself using the column names declared in `cols` using the `names` argument. Additionally, a tst file is space separated, make sure you pass ` sep=' '` to the function.

In [39]:
cols = ['time', 'rad_flow', 'fpv_close', 'fpv_open', 'high', 'bypass', 'bpv_close', 'bpv_open', 'class']
tst_url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/statlog/shuttle/shuttle.tst'

In [40]:
# your code here
shuttle = pd.read_csv(tst_url, names = cols, sep = " ")

Let's verify that this worked by looking at the `head()` function.

In [41]:
# your code here
shuttle.head()

time  rad_flow  fpv_close  fpv_open  high  bypass  bpv_close  bpv_open  \
55     0        81          0        -6    11      25         88        64   
56     0        96          0        52    -4      40         44         4   
50    -1        89         -7        50     0      39         40         2   
53     9        79          0        42    -2      25         37        12   
55     2        82          0        54    -6      26         28         2   

    class  
55      4  
56      4  
50      1  
53      4  
55      1

To make life easier for us, let's turn this dataframe into a comma separated file by saving it using the `to_csv()` function. Save `shuttle` into the file `shuttle.csv` and ensure the file is comma separated, that we are not saving the index column and that the file is saved inside the `data` folder.

In [42]:
# your code here

shuttle.to_csv("../data/shuttle.csv", index=False)

# Challenge 3 - Working with Excel Files

We can also use pandas to convert excel spreadsheets to dataframes. Let's use the `read_excel()` function. In this case, `astronauts.xls` is in the `data` folder. Read this file into a variable called `astronaut`. 

Note: Make sure to install the `xlrd` library if it is not yet installed.

In [43]:
# your code here
import xlrd 
astronaut = nasa = pd.read_excel("../data/astronauts.xls")

Use the `head()` function to inspect the dataframe.

In [44]:
# your code here
astronaut.head()

Name    Year  Group   Status Birth Date    Birth Place Gender  \
0   Joseph M. Acaba  2004.0   19.0   Active 1967-05-17  Inglewood, CA   Male   
1    Loren W. Acton     NaN    NaN  Retired 1936-03-07   Lewiston, MT   Male   
2  James C. Adamson  1984.0   10.0  Retired 1946-03-03     Warsaw, NY   Male   
3   Thomas D. Akers  1987.0   12.0  Retired 1951-05-20  St. Louis, MO   Male   
4       Buzz Aldrin  1963.0    3.0  Retired 1930-01-20  Montclair, NJ   Male   

                                          Alma Mater     Undergraduate Major  \
0  University of California-Santa Barbara; Univer...                 Geology   
1   Montana State University; University of Colorado     Engineering Physics   
2          US Military Academy; Princeton University             Engineering   
3                       University of Missouri-Rolla     Applied Mathematics   
4                           US Military Academy; MIT  Mechanical Engineering   

          Graduate Major Military Rank         Military Branch  Space Flights  \
0                Geology           NaN                     NaN              2   
1          Solar Physics           NaN                     NaN              1   
2  Aerospace Engineering       Colonel       US Army (Retired)              2   
3    Applied Mathematics       Colonel  US Air Force (Retired)              4   
4           Astronautics       Colonel  US Air Force (Retired)              2   

   Space Flight (hr)  Space Walks  Space Walks (hr)  \
0               3307            2              13.0   
1                190            0               0.0   
2                334            0               0.0   
3                814            4              29.0   
4                289            2               8.0   

                                            Missions Death Date Death Mission  
0             STS-119 (Discovery), ISS-31/32 (Soyuz)        NaT           NaN  
1                              STS 51-F (Challenger)        NaT           NaN  
2               STS-28 (Columbia), STS-43 (Atlantis)        NaT           NaN  
3  STS-41 (Discovery), STS-49 (Endeavor), STS-61 ...        NaT           NaN  
4                               Gemini 12, Apollo 11        NaT           NaN

Use the `value_counts()` function to find the most popular undergraduate major among all astronauts.

In [45]:
# your code here
astronaut_filter1= astronaut["Undergraduate Major"].value_counts()
astronaut_filter1.head(10)


Physics                     35
Aerospace Engineering       33
Mechanical Engineering      30
Aeronautical Engineering    28
Electrical Engineering      23
Engineering Science         13
Engineering                 12
Mathematics                 11
Chemistry                   10
Naval Sciences               9
Name: Undergraduate Major, dtype: int64

Due to all the commas present in the cells of this file, let's save it as a tab separated csv file. In the cell below, save `astronaut` as a **tab separated file** using the `to_csv` function. Call the file `astronaut.csv`. Remember to remove the index column and save the file in the `data` folder.

In [46]:
# your code here
# is that it? not sure if requirements have been fulfilled - what do you mean by tab separated?
astronaut.to_csv("../data/astronaut.csv", sep='\t', index=False)

# Bonus Challenge - Fertility Dataset

Visit the following [URL](https://archive.ics.uci.edu/ml/datasets/Fertility) and retrieve the dataset as well as the column headers. Determine the correct separator and read the file into a variable called `fertility`. Examine the dataframe using the `head()` function. 

Look in Google for a way to retrieve this data!

In [47]:
# your code here
# not completed
cols = ["Season", "Age", "Child_disease", "Accident/Trauma", "Fevers", "Alcohol_consumption", "Smoking", "Hours_sitting", "Output"]
fertility = pd.read_csv("../data/fertility_Diagnosis.txt", names = cols)

# index = False doens't work

fertility.head()



Season  Age  Child_disease  Accident/Trauma  Fevers  \
-0.33    0.69    0              1                1       0   
-0.33    0.94    1              0                1       0   
-0.33    0.50    1              0                0       0   
-0.33    0.75    0              1                1       0   
-0.33    0.67    1              1                0       0   

       Alcohol_consumption  Smoking  Hours_sitting Output  
-0.33                  0.8        0           0.88      N  
-0.33                  0.8        1           0.31      O  
-0.33                  1.0       -1           0.50      N  
-0.33                  1.0       -1           0.38      N  
-0.33                  0.8       -1           0.50      O

Season in which the analysis was performed. 1) winter, 2) spring, 3) Summer, 4) fall. (-1, -0.33, 0.33, 1) 

Age at the time of analysis. 18-36 (0, 1) 

Childish diseases (ie , chicken pox, measles, mumps, polio) 1) yes, 2) no. (0, 1) 

Accident or serious trauma 1) yes, 2) no. (0, 1) 

Surgical intervention 1) yes, 2) no. (0, 1) 

High fevers in the last year 1) less than three months ago, 2) more than three months ago, 3) no. (-1, 0, 1) 

Frequency of alcohol consumption 1) several times a day, 2) every day, 3) several times a week, 4) once a week, 5) hardly ever or never (0, 1) 

Smoking habit 1) never, 2) occasional 3) daily. (-1, 0, 1) 

Number of hours spent sitting per day ene-16 (0, 1) 

Output: Diagnosis normal (N), altered (O) 